File: services/witness/tests/test_canonical_leaf_v24.py
"""Cross-language parity test — Python side, v2.4.

Reads test/fixtures/canonical-leaf-v24-vectors.json and verifies the
Python twin produces byte-identical output to the TypeScript module.

Run:
    python3 services/witness/tests/test_canonical_leaf_v24.py
"""

from __future__ import annotations

import json
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent.parent))

from canonical_leaf_v24 import (  # noqa: E402
    canonical_leaf_v24,
    chain_hash_v24,
    leaf_hash_v24,
)

REPO_ROOT = Path(__file__).resolve().parents[3]
FIXTURE = REPO_ROOT / "test" / "fixtures" / "canonical-leaf-v24-vectors.json"


def main() -> int:
    with open(FIXTURE, encoding="utf-8") as f:
        fixture = json.load(f)

    failures = 0

    for v in fixture["vectors"]:
        name = v["name"]
        inp = v["input"]

        expected_preimage = v.get("expected_preimage")
        expected_hash = v.get("expected_leaf_hash")
        if expected_preimage is None or expected_hash is None:
            print(
                f"[{name}] MISSING expected values — TS side must be "
                "run with --freeze first",
                file=sys.stderr,
            )
            failures += 1
            continue

        kwargs = dict(
            tenant_id=inp.get("tenant_id", ""),
            principal_id=inp.get("principal_id", ""),
            stream_id=inp.get("stream_id", ""),
            tenant_seq=inp.get("tenant_seq", 0),
            event_time=inp.get("event_time", ""),
            payload_hash=inp.get("payload_hash", ""),
            workflow_hash=inp.get("workflow_hash", ""),
            machine_manifest_hash=inp.get("machine_manifest_hash", ""),
            dims=inp.get("dims"),
        )

        preimage = canonical_leaf_v24(**kwargs)
        computed_hash = leaf_hash_v24(**kwargs)

        if preimage != expected_preimage:
            print(f"[{name}] PREIMAGE MISMATCH", file=sys.stderr)
            print(f"  expected: {expected_preimage}", file=sys.stderr)
            print(f"  got:      {preimage}", file=sys.stderr)
            failures += 1
        if computed_hash != expected_hash:
            print(f"[{name}] LEAF_HASH MISMATCH", file=sys.stderr)
            print(f"  expected: {expected_hash}", file=sys.stderr)
            print(f"  got:      {computed_hash}", file=sys.stderr)
            failures += 1

    for c in fixture["chain_hash_vectors"]:
        name = c["name"]
        expected_chain = c.get("expected_chain_hash")
        if expected_chain is None:
            print(f"[{name}] MISSING expected_chain_hash", file=sys.stderr)
            failures += 1
            continue
        computed = chain_hash_v24(c["prev_chain_hash"], c["leaf_hash"])
        if computed != expected_chain:
            print(f"[{name}] CHAIN_HASH MISMATCH", file=sys.stderr)
            print(f"  expected: {expected_chain}", file=sys.stderr)
            print(f"  got:      {computed}", file=sys.stderr)
            failures += 1

    if failures > 0:
        print(f"\nFAIL — {failures} mismatch(es)", file=sys.stderr)
        return 1

    print(
        f"PASS — {len(fixture['vectors'])} leaf vectors + "
        f"{len(fixture['chain_hash_vectors'])} chain vectors verified."
    )
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/test-canonical-leaf-v24.ts
// Test — cross-language parity for canonicalLeafV24.
//
// Two modes:
//   * `pnpm exec tsx scripts/test-canonical-leaf-v24.ts` — verify against
//     the frozen expected values in the fixture.
//   * `pnpm exec tsx scripts/test-canonical-leaf-v24.ts --freeze` — recompute
//     and write expected values (used only on first build / rebase).
//
// See lib/witness/canonicalLeafV24.ts header for version-bump discipline.

import fs from 'node:fs';
import path from 'node:path';
import {
  canonicalLeafV24,
  leafHashV24,
  chainHashV24,
} from '../lib/witness/canonicalLeafV24';

const FIXTURE_PATH = path.join(
  process.cwd(),
  'test',
  'fixtures',
  'canonical-leaf-v24-vectors.json',
);

interface Vector {
  name: string;
  input: Record<string, unknown>;
  expected_preimage?: string;
  expected_leaf_hash?: string;
  _note?: string;
}

interface ChainVector {
  name: string;
  prev_chain_hash: string;
  leaf_hash: string;
  expected_chain_hash?: string;
  _note?: string;
}

interface Fixture {
  _notes: string[];
  vectors: Vector[];
  chain_hash_vectors: ChainVector[];
}

const mode: 'verify' | 'freeze' = process.argv.includes('--freeze')
  ? 'freeze'
  : 'verify';

const fixture: Fixture = JSON.parse(fs.readFileSync(FIXTURE_PATH, 'utf-8'));

let failures = 0;
let changed = false;

for (const v of fixture.vectors) {
  const preimage = canonicalLeafV24(v.input as never);
  const hash = leafHashV24(v.input as never);

  if (mode === 'freeze') {
    if (v.expected_preimage !== preimage || v.expected_leaf_hash !== hash) {
      v.expected_preimage = preimage;
      v.expected_leaf_hash = hash;
      changed = true;
    }
  } else {
    if (v.expected_preimage === undefined || v.expected_leaf_hash === undefined) {
      console.error(`[${v.name}] MISSING expected — run with --freeze`);
      failures++;
      continue;
    }
    if (preimage !== v.expected_preimage) {
      console.error(`[${v.name}] PREIMAGE MISMATCH`);
      console.error(`  expected: ${v.expected_preimage}`);
      console.error(`  got:      ${preimage}`);
      failures++;
    }
    if (hash !== v.expected_leaf_hash) {
      console.error(`[${v.name}] LEAF_HASH MISMATCH`);
      console.error(`  expected: ${v.expected_leaf_hash}`);
      console.error(`  got:      ${hash}`);
      failures++;
    }
  }
}

for (const c of fixture.chain_hash_vectors) {
  const computed = chainHashV24(c.prev_chain_hash, c.leaf_hash);
  if (mode === 'freeze') {
    if (c.expected_chain_hash !== computed) {
      c.expected_chain_hash = computed;
      changed = true;
    }
  } else {
    if (c.expected_chain_hash === undefined) {
      console.error(`[${c.name}] MISSING expected_chain_hash — run --freeze`);
      failures++;
      continue;
    }
    if (computed !== c.expected_chain_hash) {
      console.error(`[${c.name}] CHAIN_HASH MISMATCH`);
      console.error(`  expected: ${c.expected_chain_hash}`);
      console.error(`  got:      ${computed}`);
      failures++;
    }
  }
}

if (mode === 'freeze' && changed) {
  fs.writeFileSync(FIXTURE_PATH, JSON.stringify(fixture, null, 2) + '\n');
  console.log(`[freeze] wrote expected values to ${FIXTURE_PATH}`);
} else if (mode === 'freeze') {
  console.log('[freeze] no changes needed');
}

if (mode === 'verify') {
  if (failures > 0) {
    console.error(`\nFAIL — ${failures} mismatch(es)`);
    process.exit(1);
  }
  console.log(
    `PASS — ${fixture.vectors.length} leaf vectors + ${fixture.chain_hash_vectors.length} chain vectors verified.`,
  );
}
